embassy-hal-common/src/stm32/flash/f4.rs

```diff
@@ -0,0 +1,119 @@
+const FLASH_BASE: u32 = 0x0800_0000;
+pub(crate) const SMALL_SECTOR_SIZE: u32 = 16 * 1024;
+pub(crate) const MEDIUM_SECTOR_SIZE: u32 = 64 * 1024;
+pub(crate) const LARGE_SECTOR_SIZE: u32 = 128 * 1024;
+pub const SECOND_BANK_SECTOR_OFFSET: u8 = 12;
+
+#[derive(Debug, PartialEq)]
+pub struct FlashSector {
+    pub index: u8,
+    pub start: u32,
+    pub size: u32,
+}
+
+pub fn get_sector(address: u32, dual_bank: bool, flash_size: u32) -> FlashSector {
```

Comment:
This is hardcoding the sectors for F4/F7. Could you instead generate some static array in `build.rs` and then iterate it? This way the same code would work for all chips, including those with uniform sectors.

Something like `(size, sector_count, bank)` could work. For example for an F4 it could be:

```rust
static SECTORS: &[Sector] = &[
   Sector{ size: 16*1024, count: 4, bank: 0},
   Sector{ size: 64*1024, count: 1, bank: 0},
   Sector{ size: 128*1024, count: 3, bank: 0},
   Sector{ size: 16*1024, count: 4, bank: 1},
   Sector{ size: 64*1024, count: 1, bank: 1},
   Sector{ size: 128*1024, count: 3, bank: 1},
];
```

and for a single-bank uniform-sector chip it could be:


```rust
static SECTORS: &[Sector] = &[
   Sector{ size: 1*1024, count: 128, bank: 0},
];
```

(it's fine if the code is more complex than needed for chips with uniform sector sizes. similarly, it's fine if the `bank` stays unused for single-bank chips. We can optimize for simplicity for now, optimize later).

Reply:
This is not as easy as it sounds:) The reason is that for the F4 family, it is determined at runtime in a register setting if we are using a single or dual bank configuration. I have made an alternative proposal where we have a get_sector() per family.

I can also go with something where each family has a `get_sectors() -> &'static [Sector]` that returns the sector table depending on the the bank config, but that would require at least `#[feature(const_for)]` for a reasonable implementation.

Let me know what sounds best here...

Reply:
I found a way to do this:-)